### PR TITLE
[terra-alert]Change `customStatusColor` prop to themeable `customColorClass` prop.

### LINKED ISSUE #2307: Replace customStatusColor prop from Terra Alert
# Feature Request
Similar to: https://github.com/cerner/terra-core/issues/1999

CustomStatusColor is an untheme-able color prop and should be relplaced with a color class.

### PR DESCRIPTION
**Pull request by pranav300**
### Summary
Resolves #2307 

### Deployment Link
https://terra-core-deployed-pr-2407.herokuapp.com/#/components/terra-alert/alert/alert

### CODE REVIEW

File: packages/terra-alert/src/Alert.jsx [hunk 1]
```diff
@@ -116,13 +116,13 @@ const Alert = ({
 }) => {
   const defaultTitle = type === AlertTypes.CUSTOM ? '' : intl.formatMessage({ id: `Terra.alert.${type}` });
   const attributes = Object.assign({}, customProps);
-  const narrowAlertClassNames = cx([
+  let narrowAlertClassNames = cx([
```
Comment (nramamurth) [inline packages/terra-alert/src/Alert.jsx]: I'd directly do something like - 

```
const narrowAlertClassNames = cx([
   'alert-base',
   type,
   'narrow',
   attributes.className,
   { [`${customColorClass}`] : type === AlertTypes.CUSTOM },
]);
```

File: packages/terra-alert/docs/MoreInformation.md [hunk 2]
```diff
@@ -20,7 +20,7 @@ This image shows the layout of the elements of the Alert component.
 |action |element  |optional  |null  |An action element to be added to the action section of the alert.|
 |children  |node or string  |optional  |''  |Child Nodes providing the message content for the alert. Can contain text and HTML.|
 |customIcon  |element  |optional  |null  |The icon to be used for an alert of type custom. This will not be used for any other alert types.|
-|customStatusColor |string  |optional  |''  |The status bar color to be used for an alert of type custom.|
+|customColorClass |string  |optional  |''  |The status bar color to be used for an alert of type custom.|
```
Comment (nramamurth) [inline packages/terra-alert/docs/MoreInformation.md]: @neilpfeiffer 

From a UX standpoint, should there be a default color if this is not supplied?

File: packages/terra-alert/src/Alert.jsx (hunk 1 above)
Comment (pranav300) [reply packages/terra-alert/src/Alert.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2407/commits/fc95f5c1561b3c8f37cca2f14de071437f7cd8cf

File: packages/terra-alert/src/Alert.jsx [hunk 3]
```diff
@@ -38,7 +38,7 @@ const propTypes = {
   /**
    * The status bar color to be used for an alert of type custom.
```
[the remark this answers is not in the document]
Comment (pranav300) [reply packages/terra-alert/src/Alert.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2407/commits/a42c5b5f6e28932e9391d3ad8bbb6aa34cc9bc30.

[the remark this answers is not in the document]
Comment (ShettyAkarsh) [reply packages/terra-alert/src/Alert.jsx]: You can update the same in MoreInformation.md as well.

[the remark this answers is not in the document]
Comment (pranav300) [reply packages/terra-alert/src/Alert.jsx]: Fixed her https://github.com/cerner/terra-core/pull/2407/commits/7678ca2279c1edcf9eb0284cbde494d1e8595699.

File: packages/terra-alert/docs/MoreInformation.md [hunk 4]
```diff
@@ -20,7 +20,7 @@ This image shows the layout of the elements of the Alert component.
 |action |element  |optional  |null  |An action element to be added to the action section of the alert.|
```
Comment (jmsv6d) [inline packages/terra-alert/docs/MoreInformation.md]: I think that we could actually just remove this MoreInformation file. All this info is generated on the dev site.

File: packages/terra-alert/docs/README.md [hunk 5]
```diff
@@ -16,6 +16,12 @@ For more details, see [More Information][2].
 ```jsx
```
Comment (jmsv6d) [inline packages/terra-alert/docs/README.md]: We should remove this usage section here. The dev site allows people to see the code for the examples so this isn't needed anymore.

File: packages/terra-alert/src/Alert.module.scss [hunk 6]
```diff
@@ -122,9 +122,11 @@
     box-shadow: var(--terra-alert-success-box-shadow, 2px 0 0 #78c346 inset);
   }
 
+  /* As per the W3C spec if the box-shadow does not have the color defined, 
+  it will use the prevailing color style */
```
Comment (jmsv6d) [inline packages/terra-alert/src/Alert.module.scss]: Could you move this comment to be inline with the box-shadow style.

**Comment by ryanthemanuel** [comment]
Spacing looks weird here.  It'd be nice to fix as part of this:

image

File: packages/terra-alert/docs/README.md [hunk 7]
```diff
@@ -16,6 +16,12 @@ For more details, see [More Information][2].
...
 import React from 'react';
 import Alert from 'terra-alert';
```
Comment (BenBoersma) [inline packages/terra-alert/docs/README.md]: Probably want to remove the now unnecessary call to Alert here.

**Comment by mmalaker** [comment]
Nit: need a space between "a" and "custom" in the custom alert example.
Screen Shot 2019-06-03 at 3 39 00 PM

File: packages/terra-alert/docs/README.md (hunk 7 above)
Comment (pranav300) [reply packages/terra-alert/docs/README.md]: fixed here https://github.com/cerner/terra-core/pull/2407/commits/ffe3ff377f99567cc3e6d1d5c1be4eb714f01fe1.

File: packages/terra-alert/src/Alert.module.scss (hunk 6 above)
Comment (pranav300) [reply packages/terra-alert/src/Alert.module.scss]: fixed here https://github.com/cerner/terra-core/pull/2407/commits/ffe3ff377f99567cc3e6d1d5c1be4eb714f01fe1.

File: packages/terra-alert/docs/README.md (hunk 5 above)
Comment (pranav300) [reply packages/terra-alert/docs/README.md]: fixed here https://github.com/cerner/terra-core/pull/2407/commits/ffe3ff377f99567cc3e6d1d5c1be4eb714f01fe1.

File: packages/terra-alert/docs/MoreInformation.md (hunk 4 above)
Comment (pranav300) [reply packages/terra-alert/docs/MoreInformation.md]: fixed here https://github.com/cerner/terra-core/pull/2407/commits/ffe3ff377f99567cc3e6d1d5c1be4eb714f01fe1.

File: packages/terra-alert/src/terra-dev-site/doc/example/ActionAndDismissibleExample.jsx [hunk 8]
```diff
@@ -1,7 +1,11 @@
 import React from 'react';
 import Button from 'terra-button';
+import classNames from 'classnames/bind';
 // eslint-disable-next-line import/no-extraneous-dependencies, import/no-unresolved, import/extensions
-import Alert from 'terra-alert/lib/Alert';
+import Alert from 'terra-alert';
```
Comment (StephenEsser) [inline packages/terra-alert/src/terra-dev-site/doc/example/ActionAndDismissibleExample.jsx]: For hot reloading to work I think this needs to be reverted back to import from `'terra-alert/lib/Alert'`.

File: packages/terra-alert/tests/jest/Alert.test.jsx [hunk 9]
```diff
@@ -5,8 +5,11 @@ import { mountWithIntl } from 'terra-enzyme-intl';
 
 import IconHelp from 'terra-icon/lib/icon/IconHelp';
 import Button from 'terra-button';
+import classNames from 'classnames/bind';
 import Alert from '../../src/Alert';
+import styles from '../../src/terra-dev-site/doc/example/colors.module.scss';
 
+const cx = classNames.bind(styles);
```
Comment (StephenEsser) [inline packages/terra-alert/tests/jest/Alert.test.jsx]: This doesn't hurt anything, but it is probably overkill to use classnames to hash strings within the jest tests. The un-hashed string should be fine. The hashing can also be customized which might make this test flaky down the road.

File: packages/terra-alert/src/terra-dev-site/doc/example/ActionAndDismissibleExample.jsx (hunk 8 above)
Comment (pranav300) [reply packages/terra-alert/src/terra-dev-site/doc/example/ActionAndDismissibleExample.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2407/commits/016e1a078d9ba50f5cf9c9d7c00f95560540025f.

File: packages/terra-alert/tests/jest/Alert.test.jsx (hunk 9 above)
Comment (pranav300) [reply packages/terra-alert/tests/jest/Alert.test.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2407/commits/016e1a078d9ba50f5cf9c9d7c00f95560540025f.

File: packages/terra-alert/src/terra-dev-site/test/alert/CustomAlert.test.jsx [hunk 10]
```diff
@@ -10,14 +15,14 @@ export default () => (
...
+    <Alert id="custom4" type={Alert.Opts.Types.CUSTOM} customColorClass={cx(['terra-alert-custom-red-color'])}>Custom Alert with custom color but no title or icon</Alert>
     <br />
-    <Alert id="custom5" type={Alert.Opts.Types.CUSTOM} title="Custom_Alert:" customStatusColor="orange">Custom Alert with custom color and title but no icon</Alert>
+    <Alert id="custom5" type={Alert.Opts.Types.CUSTOM} customColorClass={cx(['terra-alert-custom-blue-color'])} title="Custom_Alert:">Custom Alert with custom color and title but no icon</Alert>
     <br />
     <Alert id="custom6" type={Alert.Opts.Types.CUSTOM} title="Custom_Alert:" customIcon={<IconHelp />}>Custom Alert with title and icon but no custom color</Alert>
     <br />
-    <Alert id="custom7" type={Alert.Opts.Types.CUSTOM} customStatusColor="orange" customIcon={<IconHelp />}>Custom Alert with custom color and icon but no title</Alert>
+    <Alert id="custom7" type={Alert.Opts.Types.CUSTOM} customColorClass={cx(['terra-alert-custom-orange-color'])} customIcon={<IconHelp />}>Custom Alert with custom color and icon but no title</Alert>
```
Comment (neilpfeiffer) [inline packages/terra-alert/src/terra-dev-site/test/alert/CustomAlert.test.jsx]:
```suggestion
    <Alert id="custom7" type={Alert.Opts.Types.CUSTOM} customColorClass={cx(['my-app-alert-dismiss-example'])} customIcon={<IconHelp />}>Custom Alert with custom color and icon but no title</Alert>
```

File: packages/terra-alert/src/terra-dev-site/test/alert/CustomAlert.test.jsx [hunk 11]
```diff
@@ -10,14 +15,14 @@ export default () => (
...
     <br />
-    <Alert id="custom8" type={Alert.Opts.Types.CUSTOM} title="Custom_Alert:" customStatusColor="orange" customIcon={<IconHelp />}>Custom Alert with title, custom color and icon</Alert>
+    <Alert id="custom8" type={Alert.Opts.Types.CUSTOM} title="Custom_Alert:" customColorClass={cx(['terra-alert-custom-blue-color'])} customIcon={<IconHelp />}>Custom Alert with title, custom color and icon</Alert>
```
Comment (neilpfeiffer) [inline packages/terra-alert/src/terra-dev-site/test/alert/CustomAlert.test.jsx]:
```suggestion
    <Alert id="custom8" type={Alert.Opts.Types.CUSTOM} title="Custom_Alert:" customColorClass={cx(['my-app-alert-help-example'])} customIcon={<IconHelp />}>Custom Alert with title, custom color and icon</Alert>
```

File: packages/terra-alert/tests/jest/Alert.test.jsx [hunk 12]
```diff
@@ -84,7 +82,7 @@ describe('Alert of type success with text content', () => {
 describe('Alert of type custom with custom title and text content', () => {
   // Snapshot Tests
   it('should render an Alert component of type custom', () => {
-    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} title="Help!" customIcon={<IconHelp />} customStatusColor="orange">This is a custom alert.</Alert>);
+    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} title="Help!" customIcon={<IconHelp />} customColorClass="terra-alert-custom-orange-color">This is a custom alert.</Alert>);
```
Comment (neilpfeiffer) [inline packages/terra-alert/tests/jest/Alert.test.jsx]:
```suggestion
    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} title="Help!" customIcon={<IconHelp />} customColorClass="my-app-alert-help-example">This is a custom alert.</Alert>);
```

File: packages/terra-alert/tests/jest/Alert.test.jsx [hunk 13]
```diff
@@ -108,7 +106,7 @@ describe('Alert of type success with an action button text content', () => {
 describe('Dismissable Alert of type custom with action button, custom title and text content', () => {
   // Snapshot Tests
   it('should render an Alert component of type custom with an action button', () => {
-    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} onDismiss={() => { }} title="Help!" customIcon={<IconHelp />} customStatusColor="orange" action={<Button text="Action" variant={Button.Opts.Variants.EMPHASIS} onClick={() => { }} />}>This is a custom alert.</Alert>);
+    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} onDismiss={() => { }} title="Help!" customIcon={<IconHelp />} customColorClass="terra-alert-custom-orange-color" action={<Button text="Action" variant={Button.Opts.Variants.EMPHASIS} onClick={() => { }} />}>This is a custom alert.</Alert>);
```
Comment (neilpfeiffer) [inline packages/terra-alert/tests/jest/Alert.test.jsx]:
```suggestion
    const wrapper = mountWithIntl(<Alert type={Alert.Opts.Types.CUSTOM} onDismiss={() => { }} title="Help!" customIcon={<IconHelp />} customColorClass="my-app-alert-help-example" action={<Button text="Action" variant={Button.Opts.Variants.EMPHASIS} onClick={() => { }} />}>This is a custom alert.</Alert>);
```

File: packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.jsx [hunk 14]
```diff
@@ -30,7 +32,7 @@ class LongTextExample extends React.Component {
         <span>
           Four score and seven years ago our fathers brought forth on this continent, a new nation, conceived in Liberty, and dedicated to the proposition that all men are created equal.
           <Toggle isOpen={!this.state.allTextShown}>
-            <Button onClick={this.handleShowMore} text="Show More" style={buttonStyle} />
+            <Button onClick={this.handleShowMore} text="Show More" className={cx('button')} />
```
Comment (neilpfeiffer) [inline packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.jsx]:
```suggestion
            <Button onClick={this.handleShowMore} text="Show More" />
```

This inline style was old and not correct - lets just remove the extra classname and have no additional styles applied (and remove the extra `LongTextExample.scss` css file). We'll fix later with an update using new `terra-show-hide` component.

File: packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.jsx [hunk 15]
```diff
@@ -39,7 +41,7 @@ class LongTextExample extends React.Component {
               But, in a larger sense, we can not dedicate -- we can not consecrate -- we can not hallow -- this ground. The brave men, living and dead, who struggled here, have consecrated it, far above our poor power to add or detract. The world will little note, nor long remember what we say here, but it can never forget what they did here. It is for us the living, rather, to be dedicated here to the unfinished work which they who fought here have thus far so nobly advanced. It is rather for us to be here dedicated to the great task remaining before us -- that from these honored dead we take increased devotion to that cause for which they gave the last full measure of devotion -- that we here highly resolve that these dead shall not have died in vain -- that this nation, under God, shall have a new birth of freedom -- and that government of the people, by the people, for the people, shall not perish from the earth.
             <br />
             <br />
-            <Button onClick={this.handleShowLess} text="Show Less" style={buttonStyle} />
+            <Button onClick={this.handleShowLess} text="Show Less" className={cx('button')} />
```
Comment (neilpfeiffer) [inline packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.jsx]:
```suggestion
            <Button onClick={this.handleShowLess} text="Show Less"  />
```
same as above - remove the extra classname

File: packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.scss [hunk 16]
```diff
@@ -0,0 +1,6 @@
+:local {
+  .button {
```
Comment (neilpfeiffer) [inline packages/terra-alert/src/terra-dev-site/doc/example/LongTextExample.scss]: Remove file

We don't need this file anymore - the inline styles that were there before were old and before when we replaced a generic button with a terra-button actually didn't apply anyway and were incorrect.

**Comment by neilpfeiffer** [comment]
+1

**Reviewed:**
- Solves for adding new customColorClass for the custom Alert type variant with appropriate documentation on how to implement.

\
\
Verified by @neilpfeiffer, Ready for Release.

**Pass** |
-----------|
UX Review|

**Pull request merged**